### PR TITLE
Add support for Commands to autodiscovery.

### PR DESCRIPTION
**Pull request by markstory**
Expand auto-discovery in CommandCollection to include Commands. Also exclude files that dont have the right suffix to prevent #11326 in the future.

Refs #11326
Refs #11137

### CODE REVIEW

**Review by robertpustulka — DISMISSED** [review]
According to the [wiki page](https://github.com/markstory/cakephp/wiki/Command-Classes#command-example) command classes should reside under `Console` namespace. We agreed on that to avoid namespace collision with command bus classes.

**Comment by markstory** [comment]
> We agreed on that to avoid namespace collision with command bus classes.

Where would core command classes go? The `src/Console` directory is going to get pretty jumbled.

File: src/Console/CommandScanner.php [hunk 1]
```diff
@@ -36,30 +36,46 @@ class CommandScanner
     public function scanAll()
     {
         $shellList = [];
-
         $appNamespace = Configure::read('App.namespace');
-        $shellList['app'] = $this->scanDir(
+
+        $coreShells = $this->scanDir(
```
Comment (ravage84) [inline src/Console/CommandScanner.php]: The method has become a little long.

How about:

```` php
    /**
     * Scan CakePHP core, the application and plugins for shell classes
     *
     * @return array A list of shell classes divided into "CORE", "app" and "plugins".
     */
    public function scanAll()
    {
        $shellClassList = [];

        $shellClassList['CORE'] = $this->_scanCore();
        $shellClassList['app'] = $this->_scanApp();;
        $shellClassList['plugins'] = $this->_scanPlugins();

        return $shellClassList;
    }

    /**
     * Scan CakePHP core for shell classes
     *
     * @return array A list of shell classes from the CakePHP core.
     */
    protected function _scanCore()
    {
        $coreShells = $this->scanDir(
            dirname(__DIR__) . DIRECTORY_SEPARATOR . 'Shell' . DIRECTORY_SEPARATOR,
            'Cake\Shell\\',
            '',
            ['command_list']
        );
        $coreCommands = $this->scanDir(
            dirname(__DIR__) . DIRECTORY_SEPARATOR . 'Command' . DIRECTORY_SEPARATOR,
            'Cake\Command\\',
            '',
            ['command_list']
        );
        $coreShellClassList = array_merge($coreShells, $coreCommands);

        return $coreShellClassList;
    }

    /**
     * Scan CakePHP app for shell classes
     *
     * @return array A list of shell classes from the app.
     */
    protected function _scanApp()
    {
        $appNamespace = Configure::read('App.namespace');

        $appShellsCLasses = $this->scanDir(
            App::path('Shell')[0],
            $appNamespace . '\Shell\\',
            '',
            ['app']
        );
        $appCommandsClasses = $this->scanDir(
            App::path('Command')[0],
            $appNamespace . '\Command\\',
            '',
            ['app']
        );
        $appShellClassList = array_merge($appShellsCLasses, $appCommandsClasses);

        return $appShellClassList;
    }

    /**
     * Scan CakePHP plugins for shell classes
     *
     * @return array A list of shell classes from the plugins.
     */
    protected function _scanPlugins()
    {
        $pluginsShellClassList = [];
        foreach (Plugin::loaded() as $plugin) {
            $path = Plugin::classPath($plugin);
            $namespace = str_replace('/', '\\', $plugin);
            $prefix = Inflector::underscore($plugin) . '.';
            $commands = $this->scanDir($path . 'Command', $namespace . '\Command\\', $prefix, []);
            $shells = $this->scanDir($path . 'Shell', $namespace . '\Shell\\', $prefix, []);
            $pluginsShellClassList[$plugin] = array_merge($shells, $commands);
        }

        return $pluginsShellClassList;
    }
````

Comment (markstory) [reply src/Console/CommandScanner.php]: Good idea!

**Pull request merged**